File: sections/65_polynomial_algorithms_recurrences/problems/b_bounded_sum_product/solution.py
import heapq
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))
from ntt import convolution


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, target = data[:2]
    heap: list[tuple[int, int, list[int]]] = []
    serial = 0
    for bound in data[2 : 2 + n]:
        if bound > 0:
            polynomial = [1] * (min(bound, target) + 1)
            heapq.heappush(heap, (len(polynomial), serial, polynomial))
            serial += 1
    if not heap:
        print(int(target == 0))
        return
    while len(heap) > 1:
        _, _, left = heapq.heappop(heap)
        _, _, right = heapq.heappop(heap)
        product = convolution(left, right, target + 1)
        heapq.heappush(heap, (len(product), serial, product))
        serial += 1
    answer = heap[0][2]
    print(answer[target] if target < len(answer) else 0)


if __name__ == "__main__":
    main()

File: sections/65_polynomial_algorithms_recurrences/problems/b_bounded_sum_product/solution.cpp
#include <bits/stdc++.h>
#include "../../ntt.hpp"
using namespace std;
using course_ntt::convolution;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, target;
    if (!(cin >> n >> target)) return 0;

    struct Compare {
        bool operator()(const vector<int> &a, const vector<int> &b) const {
            return a.size() > b.size();
        }
    };
    priority_queue<vector<int>, vector<vector<int>>, Compare> products;
    for (int i = 0, bound; i < n; ++i) {
        cin >> bound;
        if (bound > 0) products.push(vector<int>(min(bound, target) + 1, 1));
    }
    if (products.empty()) {
        cout << (target == 0) << '\n';
        return 0;
    }
    while (products.size() > 1) {
        auto left = products.top(); products.pop();
        auto right = products.top(); products.pop();
        products.push(convolution(left, right, target + 1));
    }
    const auto &answer = products.top();
    cout << (target < static_cast<int>(answer.size()) ? answer[target] : 0) << '\n';
}
